### PR TITLE
Extended authentication

### LINKED ISSUE #637: AWS ECR login credential exchange
[AWS ECR](http://docs.aws.amazon.com/AmazonECR/latest/userguide/Registries.html) requires a [cli login](http://docs.aws.amazon.com/cli/latest/reference/ecr/get-login.html) before any docker image push or pull.  I currently have my jenkins job running the `aws ecr get-login` command.  However, I prefer not install the amazon toolkit on all my build infrastructure and require all our developers to also install the amazon toolkit.

I would like to submit a PR that would invoke the [GetAuthorizationToken api](http://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html) to exchange AWS IAM credentials for temporary ECR credentials.  This PR would augment the AuthConfigFactory.  Any login credentials specified for an ECR repository would be considered IAM credentials needing exchange.

An ECR repository has a URL of form **https://**_awsAccountId_**.dkr.ecr.**_awsRegion_**.amazonaws.com**/_repositoryName_

1. Should credentials explicitly specified with docker.* properties be exchanged?
2. Should credentials from ~/.config/kube be exchanged?
3. Should credentials provided in plugin configuration be exchanged?
4. Should credentials from .m2/settings be exchanged?
5. Should credentials from ~/.docker/config.json be exchanged?
6. Should there be a flag to prevent the exchange?

I'm thinking that credentials 1, 2, 3, 4 should be exchanged.  ~/.docker/config.json likely contains already exchanged credentials. ~/.config/kube likely contains IAM credentials, as kubernetes ["will fetch and periodically refresh ECR credentials"](http://kubernetes.io/docs/user-guide/images/#using-aws-ec2-container-registry) .

### PR DESCRIPTION
**Pull request by chonton**
AWS ECR key exchange
closes #637
replaces PR #646

### CODE REVIEW

**Comment by rhuss** [comment]
Thanks !

**Comment by chonton** [comment]
Working on unit tests to extend coverage

File: src/main/java/io/fabric8/maven/docker/access/ecr/AwsSigner4Request.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            signed.append(header.getKey());
+
+            squeezeWhite(canonical, header.getKey());
+            canonical.append(':');
+            squeezeWhite(canonical, header.getValue().toString());
+            canonical.append('\n');
+        }
+    }
+
+    private static void squeezeWhite(StringBuilder dst, String src) {
```
Comment (chonton) [inline src/main/java/io/fabric8/maven/docker/access/ecr/AwsSigner4Request.java]: done

File: src/main/java/io/fabric8/maven/docker/access/ecr/AwsSigner4Request.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    private static void canonicalizeHeaders(HttpRequest request, StringBuilder canonical, StringBuilder signed) {
+        Map<String, StringBuilder> unique = new TreeMap<>();
+        for (Header header : request.getAllHeaders()) {
+            String key = header.getName().toLowerCase(Locale.US);
+            if (key.equals("connection")) {
+                continue;
+            }
+            StringBuilder builder = unique.get(key);
+            if (builder != null) {
+                if (builder.length() > 0) {
```
Comment (chonton) [inline src/main/java/io/fabric8/maven/docker/access/ecr/AwsSigner4Request.java]: refactored this method.  complicated by fact it was trying to accomplish two tasks in same loop.

File: src/main/java/io/fabric8/maven/docker/util/AuthConfigFactory.java [hunk 3]
```diff
@@ -44,19 +47,111 @@
...
+     * @param user user to check for
+     * @param registry registry to use, might be null in which case a default registry is checked,
+     * @return the authentication configuration or <code>null</code> if none could be found
+     *
+     * @throws MojoFailureException
+     */
+    public AuthConfig createAuthConfig(boolean isPush, boolean skipExtendedAuth, Map authConfig, Settings settings, String user, String registry)
+            throws MojoExecutionException {
+
+        AuthConfig ret = createAuthConfig(isPush, authConfig, settings, user, registry);
```
Comment (chonton) [inline src/main/java/io/fabric8/maven/docker/util/AuthConfigFactory.java]: done

**Comment by chonton** [comment]
Ping. Any additional changes needed?

**Comment by rhuss** [comment]
Just back from holidays, I will have a look today.

**Comment by rhuss** [comment]
[merge]

**Pull request merged**